com.sap.cloud.lm.sl.cf.web/src/main/java/com/sap/cloud/lm/sl/cf/web/security/RateLimiterFilter.java

```diff
@@ -0,0 +1,64 @@
+package com.sap.cloud.lm.sl.cf.web.security;
+
+import com.sap.cloud.lm.sl.cf.web.Constants;
+import com.sap.cloud.lm.sl.cf.web.helpers.RateLimiterProvider;
+import io.github.resilience4j.ratelimiter.RateLimiter;
+import io.github.resilience4j.ratelimiter.RateLimiterConfig;
+import io.github.resilience4j.ratelimiter.internal.AtomicRateLimiter.AtomicRateLimiterMetrics;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
+
+@Named("rateLimiter")
+public class RateLimiterFilter extends OncePerRequestFilter {
+
+    @Inject
+    private RateLimiterProvider rateLimiterProvider;
+
+    @Override
+    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain)
+        throws ServletException, IOException {
+        String ipAddress = request.getRemoteAddr();
+        RateLimiter rateLimiter = rateLimiterProvider.getRateLimiter(ipAddress);
+
+        boolean hasAcquiredPermission = rateLimiter.acquirePermission();
+        AtomicRateLimiterMetrics metrics = (AtomicRateLimiterMetrics) rateLimiter.getMetrics();
+        RateLimiterConfig config = rateLimiter.getRateLimiterConfig();
+
+        response.setHeader(Constants.RATE_LIMIT, Integer.toString(config.getLimitForPeriod()));
+        response.setHeader(Constants.RATE_LIMIT_REMAINING, Integer.toString(metrics.getAvailablePermissions()));
+        response.setHeader(Constants.RATE_LIMIT_RESET, Long.toString(getUtcTimeForNextReset(config, metrics)));
+
+        if (!hasAcquiredPermission) {
+            response.sendError(HttpStatus.TOO_MANY_REQUESTS.value(), HttpStatus.TOO_MANY_REQUESTS.getReasonPhrase());
+            return;
+        }
+        filterChain.doFilter(request, response);
+    }
+
+    private static long getUtcTimeForNextReset(RateLimiterConfig config, AtomicRateLimiterMetrics metrics) {
+        long cyclePeriodInNanos = config.getLimitRefreshPeriod()
+                                        .toNanos();
+        long currentCycle = metrics.getCycle();
+        long nanosToWaitForReset = (currentCycle + 1) * cyclePeriodInNanos;
```

Comment:
Why incrementing the current cycle?
If I understood correctly, the whole calculation would do this:
1. Get the currentCycle (let's say 5th cycle)
2. nanosToWaitForReset = 6 * 1hour as nanos
3. now + nanosToWaitForReset = 6 cycles ahead
This means that value returned from this method will be the release after six cycles, from now.
Have I understood something wrong?

Reply:
Is there some kind of a method which is saying how many nano seconds have been passed since the start of the current cycle?
There is a method in the AtomicRateLimiterMetrics:
https://github.com/resilience4j/resilience4j/blob/master/resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/AtomicRateLimiter.java#L413
Isn't that going to return the value which need to be waited before the next cycle?